### DISCUSSION TITLE
Brand data functions, dealing with < 9.6.0, taxonomy clashes, migration of brand attribute terms

### DISCUSSION THREAD

**Discussion by medavidallsop** [woocommerce/woocommerce#52774] [Brands]
Hi,

I have been reading https://developer.woocommerce.com/2024/10/01/introducing-brands/ and can see this is an opt-in feature, and will be enabled for all users in 9.6 (very soon).

I have a few discussion points:

1. How do you get brands data, will the WC_Product object be updated to include things like get_brands (to get brands associated with a product), will there be a new overall function like get_brands (to get all brands)
2. Can you provide recommendations on how developers should condition functionality based on the above, really basic example: You have a plugin that adds the name(s) of brands associated with a product to the product page, in the code that plugin gets the brands data from the WC_Product object to get the brands associated and loops through them to display them - however what if the user has the latest version of this example plugin but a < 9.6.0 Woo version, the functions used to get brand data will probably be fatal errors as those functions are yet to exist, therefore what is the recommendation for dealing with this scenario
3. I assume the brands is just a product taxonomy, are there any potential taxonomy key clash issues here with other plugins that may have created a taxonomy called the same as the core brand taxonomy (I haven't checked what it is, but i'm guessing it will be brand or something quite generic)
4. Before this functionality / without using the separate plugin, tons of websites assign a brand via product attribute terms, is there any plans to help migrate that data somehow to this new method?

Thank you,

**Comment by jimjasson** [comment]
Hey @medavidallsop, 

Thank you for the questions! 

1) 

You may get the brands that are assigned to a specific product by running `wc_get_brands`: https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/wc-brands-functions.php#L85C10-L85C23

To get all brands from a store, you may run a `WP_Query` using the `product_brand` taxonomy, similar to how you retrieve all product categories. 

2) 

> however what if the user has the latest version of this example plugin but a < 9.6.0 Woo version, the functions used to get brand data will probably be fatal errors as those functions are yet to exist, therefore what is the recommendation for dealing with this scenario

If the Brands module is not activated on a store or WooCommerce less than 9.6.0 is used, then you are right that brands-related functions will not be accessible. For that reason, any plugin that integrates with the brands functionality offered by WooCommerce core must require WooCommerce 9.6.0 as the minimum required version or disable some features for stores using earlier versions. 

Keep in mind that all the brands logic that is now part of core was previously offered by the [WooCommerce Brands](https://github.com/woocommerce/woocommerce-brands) plugin. If a plugin integrated with WooCommerce Brands, then this integration should continue to work after WooCommerce 9.6.0, as well. 

3) 

If another plugin uses the same `product_brand` taxonomy or any function used by WooCommerce core, then indeed this can lead to a fatal error. The same is true for all taxonomies defined by WooCommerce core, like product categories, or any functions that exist in core. 

That said, we have already tested the most popular brands plugins and confirmed that there isn't any conflict in regards to taxonomies or function names -- here's a list for reference:
- Brands for WooCommerce 
- MAS Brands for WooCommerce 
- Perfect Brands for WooCommerce
- Premmerce Brands for WooCommerce 
- Smart Brands for WooCommerce
- WPC Brands for WooCommerce 
- YITH WooCommerce Brands Add-On 
- Ultimate WooCommerce Brands
- Easy Woocommerce Brands 
- WSB Brands

4) 

As I mentioned in (2), all the brands logic that is now part of core was previously offered by the [WooCommerce Brands](https://github.com/woocommerce/woocommerce-brands) plugin. Users of this plugin will automatically see all their brands migrated to core without any further action, as that the same taxonomy is used. If a store created brands using a custom taxonomy, attribute or any other way, then this would require custom work, as we cannot account for all the custom brands implementations that have been created.

**Reply by dhilditch** [reply]
re: 3 it seems you never bothered about product attributes. It would have been nice if you had added some check to see if something else is using /brand/ as a qualifier, e.g. many of my customers who never bothered with a brand plugin and just used an attribute for brand.

**Reply by jimjasson** [reply]
Hey @dhilditch, 

Having chatted with a lot of folks after the release of WooCommerce v9.6, which enabled brands for all users, I confirmed our assumption that being able to organize products by brand is a very common merchandizing need and it makes sense for WooCommerce to offer it for free. Since WooCommerce didn't offer this feature out of the box for so long, though, many merchants either used an extension that offered this functionality, a custom attribute or another custom solution. 

I understand that if you have been using a different brands solution in your store, it now seems a lot of extra work to migrate to WooCommerce core brands feature. However, going forward, merchants who didn't have brands in their store can now start creating some without the overhead of searching for an additional plugin or creating a custom solution. 

If you are looking to migrate from your custom solution to the core's brands feature, then the migration process depends on your current setup. 

If you were using a plugin that created a brands taxonomy, there are [some plug and play solutions](https://github.com/woocommerce/woocommerce/discussions/54720#discussioncomment-11926622) that you can use to migrate from the plugin's brands taxonomy to core's taxonomy. 

If you were using a custom attribute (which seems to be the case), then a WooCommerce community developer shared [this migration snippet in another thread](https://github.com/woocommerce/woocommerce/discussions/54728#discussioncomment-12040006). 

If you decide to try out one of the above options, I encourage you to do so on a staging site first. Depending on how many brands/products you currently have, you might need to make some optimizations in order for the migration process to run in batches. 

That said, you can continue using your own custom implementation of brands until you find some time to explore the migration options further. In the meantime, you may disable the Brands module by running this WP CLI command: `wp option update wc_feature_woocommerce_brands_enabled 'no'`.

If you notice any conflicts regarding the use of the `brands` slug, then, after you disable the Brands module, I recommend regenerating your site's permalinks by going to the Settings > Permalinks tab and clicking on 'Save changes' (you don't have to make any changes to click Save changes).

**Comment by medavidallsop** [comment]
Thanks for the info.

For 1: Would there be any intention of including brand function(s) like get brand, set brand, etc via the WC_Product object? Once this becomes part of core WooCommerce it would make sense for this data to be got/set, etc via the WC_Product object like all other product data can? + leaving the earlier stuff for wc_get_brands available for compatibility?

**Reply by jimjasson** [reply]
Good idea @medavidallsop! 

This is definitely something to consider after WooCommerce 9.6 is released, when Brands will be enabled for all users. I created an [issue](https://github.com/woocommerce/woocommerce/issues/52991) to track progress on this.
